// ng2-swipe-cards/esm/ng2-swipe-cards.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CardComponent } from './components/card';
import { TinderCardDirective } from './directives/tinder-card';
import { HammerGestureConfig, HAMMER_GESTURE_CONFIG } from '@angular/platform-browser';
export class HammerConfig extends HammerGestureConfig {
    constructor() {
        super(...arguments);
        this.overrides = {
            'pan': { enable: true }
        };
    }
}
export class SwipeCardsModule {
}
SwipeCardsModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule
                ],
                declarations: [
                    CardComponent,
                    TinderCardDirective
                ],
                exports: [
                    CardComponent,
                    TinderCardDirective
                ],
                providers: [{
                        provide: HAMMER_GESTURE_CONFIG,
                        useClass: HammerConfig
                    }]
            },] },
];
/** @nocollapse */
SwipeCardsModule.ctorParameters = () => [];
export { CardComponent } from './components/card';
export { TinderCardDirective } from './directives/tinder-card';
